sopel/modules/currency.py

```diff
@@ -1,102 +1,189 @@
 # coding=utf-8
 # Copyright 2013 Elsie Powell, embolalia.com
+# Copyright 2019 Mikkel Jeppesen
 # Licensed under the Eiffel Forum License 2
 from __future__ import unicode_literals, absolute_import, print_function, division
 
+import time
 import re
+import requests
 
-from requests import get
-from sopel.module import commands, example, NOLIMIT
+from sopel.module import commands, example, NOLIMIT, rule
+from sopel.tools import stderr
+from sopel.config.types import StaticSection, ValidatedAttribute
 
-# The Canadian central bank has better exchange rate data than the Fed, the
-# Bank of England, or the European Central Bank. Who knew?
-base_url = 'https://www.bankofcanada.ca/valet/observations/FX{}CAD/json'
+rates_fiat_json = {}
+rates_btc_json = {}
+
+
+fiat_url = 'https://api.exchangeratesapi.io/latest?base=EUR'
+fixer_url = 'http://data.fixer.io/api/latest?base=EUR&access_key={}'
+crypto_url = 'https://apiv2.bitcoinaverage.com/indices/global/ticker/short?crypto=BTC'
 regex = re.compile(r'''
-    (\d+(?:\.\d+)?)        # Decimal number
-    \s*([a-zA-Z]{3})       # 3-letter currency code
-    \s+(?:in|as|of|to)\s+  # preposition
-    ([a-zA-Z]{3})          # 3-letter currency code
-    ''', re.VERBOSE)
-
-
-def get_rate(code):
-    code = code.upper()
-    if code == 'CAD':
-        return 1, 'Canadian Dollar'
-    elif code == 'BTC':
-        btc_rate = get('https://apiv2.bitcoinaverage.com/indices/global/ticker/BTCCAD')
-        rates = btc_rate.json()
-        return 1 / rates['averages']['day'], 'Bitcoin—24hr average'
-
-    data = get(base_url.format(code))
-    name = data.json()['seriesDetail']['FX{}CAD'.format(code)]['description']
-    name = name.split(" to Canadian")[0]
-    json = data.json()['observations']
-    for element in reversed(json):
-        if 'v' in element['FX{}CAD'.format(code)]:
-            return 1 / float(element['FX{}CAD'.format(code)]['v']), name
-
-
-@commands('cur', 'currency', 'exchange')
-@example('.cur 20 EUR in USD')
-def exchange(bot, trigger):
+    ^(\d+(?:\.\d+)?)                                            # Decimal number
+    \s*([a-zA-Z]{3})                                            # 3-letter currency code
+    \s+(?:in|as|of|to)\s+                                       # preposition
+    (([a-zA-Z]{3}$)|([a-zA-Z]{3})\s)+$                          # one or more 3-letter currency code
+''', re.VERBOSE)
+
+
+def btc_rate(code, reverse=False):
+    global rates_btc_json
```

Comment:
This line `global ...` is not needed since rates_btc_json is not being modified here.

Reply:
removed :)